### PR TITLE
Feature boundary condition

### PR DESCRIPTION
**Pull request by PhilippeBussetta**
Regarding the conversation  #1346, I have added the possibility to assign value or vector versus time interval

Tests in release as well as debug: OK

```
runkratos.exe kratos\python_scripts\run_tests.py -a StructuralMechanicsApplication -l nightly
 |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 6.0.0-5e69c913ca-FullDebug
Running tests for KratosCore
D:\Kratos-FeatureElement/kratos/tests/test_KratosCore.py
.............x..s.....................................s...s.............s...........................................ss
----------------------------------------------------------------------
Ran 118 tests in 184.546s

OK (skipped=6, expected failures=1)
Running tests for StructuralMechanicsApplication
D:\Kratos-FeatureElement\applications//StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py
................................................................................................................................
----------------------------------------------------------------------
Ran 128 tests in 483.944s

OK
Running cpp tests
..............................................................................................................................................................................................................................................................................................................
Ran 302 of 303 test cases in 10.5387s. OK.
```

### CODE REVIEW

**Review by loumalouomega — COMMENTED** [review]
Right now it is not necessary if we can consider any analytical function, what do you thing @RiccardoRossi ?

**Comment by PhilippeBussetta** [comment]
It can be useful to consider analytical function or piecewise function.
Ex: linear if t< t1 and constant if t >= t1

**Comment by PhilippeBussetta** [comment]
Ping @RiccardoRossi @loumalouomega

**Comment by PhilippeBussetta** [comment]
Ping @RiccardoRossi @loumalouomega

**Comment by loumalouomega** [comment]
I prefer to wait to @RiccardoRossi, the definition of these processes is a very sensible matter for him

**Comment by RiccardoRossi** [comment]
Hi @PhilippeBussetta the original design idea for this kind of use was to use two processes each acting in the corresponding interval. Clearly if the intervals are consecutive you would get the same behaviour as you are obtaining.
What would be the problem with doing that?
How is the new format you are proposing? 

Note also that if the new behaviour is accepted a test of the new behaviour should be added in the test_processes, ideally testing corner cases as time exactly coinciding to the interval limits)

**Comment by PhilippeBussetta** [comment]
Hi @RiccardoRossi the problem with the 2 processes way is the overlapping or gap because of numerical approximation.

With the new format:
-the time interval can subdivided into sub-interval
-the values can be applied over interval -- as previously, number or function of time and space -- or at node with interpolation function over the interval -- only numerical value.
-the previous format is still working

Of course, if the new format is accepted, I will add a test.

**Comment by RiccardoRossi** [comment]
i @PhilippeBussetta the current processes are in principle designed so that you can concatenate them. this means that in principle they are acting in the interval    tbegin <= t < tend
this is actually the reason for which the interval was centralized in a function.

Having said this i wouldn't say this feature was extensively tested, so i don't exclude failures. 
If it did work correctly, would the solution i am proposing be ok for you?

**Comment by PhilippeBussetta** [comment]
Hi, @RiccardoRossi , I don't understand your solution.
The solution of this PR could be split into 2:
-definition of one process with multi-time interval -- merged processes.
-definition of the assigned value at time value with the name of an interpolation function.

On my point of view the 1st point is more reliable than the multi-process way and reduce the json file size.
the 2nd, is just to reduce the expression of the function into the json file.

Maybe we can create 2 process or just remove the 2nd part.

**Comment by RiccardoRossi** [comment]
@PhilippeBussetta i guess i don't understand how your process looks like from the input point of view.

could you copy paste here the definition of your new process as it would appear in the ProjectParameters.json?

**Comment by PhilippeBussetta** [comment]
An example:
```
        {
            "python_module": "assign_vector_variable_process",
            "kratos_module": "KratosMultiphysics",
            "help": "This process fixes the selected components of a given vector variable",
            "process_name": "AssignVectorVariableProcess",
            "Parameters": {
                "model_part_name": "Part_Displacement_Max",
                "mesh_id": 0,
                "variable_name": "DISPLACEMENT",
                "constrained": [true, true, true],
                "value": [ [0.0, 12.6, 12.6], 0.0, 0.0],
                "interval": [ 0.0, 10.0, "End"],
                "step_type": "linear"
            }
        }
```
or 

```
        {
            "python_module": "assign_vector_variable_process",
            "kratos_module": "KratosMultiphysics",
            "help": "This process fixes the selected components of a given vector variable",
            "process_name": "AssignVectorVariableProcess",
            "Parameters": {
                "model_part_name": "Part_Displacement_Max",
                "mesh_id": 0,
                "variable_name": "DISPLACEMENT",
                "constrained": [true, true, true],
                "value": [ [12.6*t, 126.0], 0.0, 0.0],
                "interval": [ 0.0, 10.0, "End"],
                "step_type": ""
            }
        }
```

**Comment by RiccardoRossi** [comment]
Hi @PhilippeBussetta i am not 100% sure i understand your syntax, but my guess is that it would be equivalent to

    "list_of_processes" : [
       ....,
       {
            "python_module": "assign_vector_variable_process",
            "process_name": "AssignVectorVariableProcess",
            "Parameters": {
                "model_part_name": "Part_Displacement_Max",
                "variable_name": "DISPLACEMENT",
                "constrained": true,
                "value": [ 12.6*t, 0.0, 0.0],
                "interval": [ 0.0, 10.0],
            }
        },
       {
            "python_module": "assign_vector_variable_process",
            "process_name": "AssignVectorVariableProcess",
            "Parameters": {
                "model_part_name": "Part_Displacement_Max",
                "variable_name": "DISPLACEMENT",
                "constrained": true,
                "value": [ 12.6, 0.0, 0.0],
                "interval": [ 10.0, "End"],
            }
        }
     ]
   
where i took out all of the unnecesary items to keep it as short as possible.
if this is so **my personal taste** would be for this second option. 
The techincal rationale is that implementation is easier, and i would also say it is more readable (although admittedly longer)

Still, since i am speaking of **personal taste** what do other people think? @KratosMultiphysics/technical-committee ??

**Comment by philbucher** [comment]
for me the version proposed by @RiccardoRossi is much more readable
I can understand right away what is applied where

@PhilippeBussetta I dont understand what "step_type" means. Also I don't understand how/when "value" is being used

Therefore the proposal of Riccardo makes more sense to me, it seems more flexible 
I guess there would be a solution for the numerical errors at the end one one/beginning of another step

**Comment by PhilippeBussetta** [comment]
Then, what we do?
No more opinion?
I delete the modifications?
With this PR it will be possible to used the old syntax (see example or @RiccardoRossi)
I add another possibility.

So, we merge the PR?

Ping @KratosMultiphysics/technical-committee

**Comment by philbucher** [comment]
@PhilippeBussetta if this is still relevant could you please give an answer to the questions I asked in my last comment?

**Comment by PhilippeBussetta** [comment]
I am sorry, @philbucher  

"step_type" is the choice of the interpolation method.
The first example use "linear" that means that the value into the interval is the linear interpolation of the values at the boundaries.
A "smooth" method is also implemented -- hyperbolic tan function. [here](https://github.com/KratosMultiphysics/Kratos/pull/1930/files#diff-9af79811c5e1500e6e60d11b1ec66947R118)

"step_type" is not needed in the second example

**Comment by philbucher** [comment]
@PhilippeBussetta thanks for the explanation

I looked through the code again and my comments are still the same as what I wrote on the 30th of April :)
Except that now I know what step_type means. 

Regarding the step type: This is (as I understand) conflicting with the definition of a "function" as a value
=> If I specify a function it can be anything (linear, hyperbolic, whatever). Thus defining the step_type seems redundant to me.
What happens if the user specifies both?

As @RiccardoRossi was asking, would his solution (his last post of 30th April) be acceptable for you given that it works properly btw the intervals?

Edit: with your proposed implementation it is also not possible to change the constraint inbetween the intervals, which would be possible with using two separate processes

**Comment by PhilippeBussetta** [comment]
The 'step_type' is not redundant, it is another option.
If the user specify a function -- one value corresponding to 2 times (one interval), the 'step_type' is not used.
We can add a error if the 'step_type' is specified.
If the user specify one value for each time -- 2 values for an simple interval -- the 'step_type' have to be specified.

@philbucher I don't understand your last comment. With this implementation is still possible to use two separate processes.

**Comment by philbucher** [comment]
@PhilippeBussetta regarding step_type, since the same can be achieved specifying a function (you showed this yourself in your example!) I would say it is redundant.

Regarding my last comment quoting you:
> Hi @RiccardoRossi the problem with the 2 processes way is the overlapping or gap because of numerical approximation.

If the current implementation is improved such that the overlapping is working properly, would this be an acceptable solution for you?

**Comment by PhilippeBussetta** [comment]
Yes, nevertheless my process is 2x shorter regarding the json file.

**Comment by philbucher** [comment]
But the implementation is, at least imo, more complicated

eventually it is the decision of @RiccardoRossi and @KratosMultiphysics/technical-committee , I pointed out my concerns

**Comment by PhilippeBussetta** [comment]
Only a little bit more complex ;-)
Of course, it is the decision of @RiccardoRossi and @KratosMultiphysics/technical-committee

**Comment by PhilippeBussetta** [comment]
I remove the new syntax of the assign function.
I add variable in the shell element.

File: applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp [hunk 1]
```diff
@@ -387,8 +387,9 @@ void ShellThinElement3D4N::GetValueOnIntegrationPoints
     {
         caseId = 30;
     }
-
-
+    else if (rVariable == SHEAR_ANGLE) {
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp]: This has nothing to do with this PR
Please make a separate PR for this feature

File: applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp [hunk 2]
```diff
@@ -463,6 +464,11 @@ void ShellThinElement3D4N::GetValueOnIntegrationPoints
 
                 CalculateVonMisesStress(data, rVariable, resultDouble);
             }
+            else if (caseId == 40) {
+                DecimalCorrection(data.generalizedStrains);
+                const double pi = 3.14159265;
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp]: use Kratos::Globals, don't declare this yourself

File: applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp [hunk 3]
```diff
@@ -463,6 +464,11 @@ void ShellThinElement3D4N::GetValueOnIntegrationPoints
 
                 CalculateVonMisesStress(data, rVariable, resultDouble);
             }
+            else if (caseId == 40) {
+                DecimalCorrection(data.generalizedStrains);
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D4N.cpp]: Why do you use this?

**Review by philbucher — DISMISSED** [review]
Please do not mix features, make a separate PR for different things
After doing this I have some comments for the shells, please put me in review

**Comment by PhilippeBussetta** [comment]
I split it

**Comment by PhilippeBussetta** [comment]
Could you valid it, I just remove duplicated code

File: kratos/python_scripts/assign_vector_variable_process.py [hunk 4]
```diff
@@ -35,17 +35,8 @@ def __init__(self, Model, settings ):
                 for i in range(3):
                     settings["constrained"][i].SetBool(is_fixed)
 
-
-        #detect "End" as a tag and replace it by a large number
-        if(settings.Has("interval")):
-            if(settings["interval"][1].IsString() ):
-                if(settings["interval"][1].GetString() == "End"):
```
Comment (philbucher) [inline kratos/python_scripts/assign_vector_variable_process.py:42]: Why do you remove this, I don't think this is duplicated

Comment (PhilippeBussetta) [reply kratos/python_scripts/assign_vector_variable_process.py:42]: It is implemented in IntervalUtility and called though the assign_scalar-variable_process.py

**Comment by PhilippeBussetta** [comment]
ping

File: kratos/python_scripts/assign_vector_variable_process.py (hunk 4 above)
Comment (philbucher) [reply kratos/python_scripts/assign_vector_variable_process.py:42]: yep I see

File: kratos/python_scripts/assign_vector_variable_process.py [hunk 5]
```diff
@@ -57,41 +48,18 @@ def __init__(self, Model, settings ):
...
-            z_params.AddValue("model_part_name",settings["model_part_name"])
-            z_params.AddValue("mesh_id",settings["mesh_id"])
-            z_params.AddEmptyValue("constrained").SetBool(settings["constrained"][2].GetBool())
-            z_params.AddValue("interval",settings["interval"])
-            z_params.AddValue("value",settings["value"][2])
-            z_params.AddEmptyValue("variable_name").SetString(settings["variable_name"].GetString() + "_Z")
-            z_params.AddValue("local_axes",settings["local_axes"])
-            self.aux_processes.append( assign_scalar_variable_process.AssignScalarVariableProcess(Model, z_params) )
+        # loop over components X, Y and Z
+        for variable_i in enumerate(["_X", "_Y", "_Z"]):
```
Comment (philbucher) [inline kratos/python_scripts/assign_vector_variable_process.py]: last question, why do you need enumerate?
doesn't `for variable_i in ["_X", "_Y", "_Z"]:` also work?

Comment (philbucher) [reply kratos/python_scripts/assign_vector_variable_process.py]: I think I understand now, but it is not straight forward, why don't you use sth like
`for variable_i, variable_extension in enumerate(["_X", "_Y", "_Z"])`, 
then you don't have to do the nested list manipulation

Comment (PhilippeBussetta) [reply kratos/python_scripts/assign_vector_variable_process.py]: Just to reduce the size, write less
On my point of view, it is equal

**Review by philbucher — DISMISSED** [review]
approving since its only cleanup

**Review by philbucher — APPROVED** [review]
